File: task-module/src/main/java/com/pebble/task/config/SecurityConfig.java
package com.pebble.task.config;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.http.SessionCreationPolicy;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.AuthenticationEntryPoint;

@Configuration
@EnableWebSecurity
public class SecurityConfig {

    @Bean
    public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
        http
            .csrf(csrf -> csrf.disable())
            .sessionManagement(session -> session.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
            .authorizeHttpRequests(auth -> auth
                .requestMatchers("/actuator/health", "/", "/index.html", "/static/**").permitAll()
                .anyRequest().authenticated()
            )
            .oauth2ResourceServer(oauth2 -> oauth2
                .jwt(jwt -> {})
                .authenticationEntryPoint(jsonAuthenticationEntryPoint()) // JSON 응답 설정
            );

        http.headers(headers -> headers.frameOptions(frame -> frame.disable()));
        return http.build();
    }

    @Bean
    public AuthenticationEntryPoint jsonAuthenticationEntryPoint() {
        return (request, response, authException) -> {
            response.setStatus(HttpStatus.UNAUTHORIZED.value());
            response.setContentType(MediaType.APPLICATION_JSON_VALUE);
            response.setCharacterEncoding("UTF-8");
            String json = "{\"status\": 401, \"error\": \"Unauthorized\", \"message\": \"인증이 필요한 서비스입니다. 게이트웨이 포털에서 로그인을 먼저 진행해주세요.\", \"path\": \"" + request.getRequestURI() + "\"}";
            response.getWriter().write(json);
        };
    }
}

File: auth-module/src/main/kotlin/com/pebble/basicAuth/config/SecurityConfig.kt
package com.pebble.basicAuth.config

import com.pebble.basicAuth.config.oauth2.CustomOAuth2UserService
import com.pebble.basicAuth.config.oauth2.OAuth2SuccessHandler
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.security.web.util.matcher.AntPathRequestMatcher

@Configuration
@EnableMethodSecurity
class SecurityConfig(
    private val authenticationHandler: CustomAuthenticationHandler,
    private val jwtAuthenticationFilter: JwtAuthenticationFilter,
    private val customOAuth2UserService: CustomOAuth2UserService,
    private val oAuth2SuccessHandler: OAuth2SuccessHandler,
    private val formLoginSuccessHandler: FormLoginSuccessHandler,
    private val formLoginFailureHandler: FormLoginFailureHandler
) {

    @Bean
    @Throws(Exception::class)
    fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
        http
            .csrf { it.disable() } // 테스트 편의를 위해 CSRF 비활성화
            
            // 1. 세션 정책: 인증 서버는 로그인 과정을 위해 세션이 필요함 (IF_REQUIRED)
            .sessionManagement { session ->
                session.sessionCreationPolicy(SessionCreationPolicy.IF_REQUIRED)
            }

            .authorizeHttpRequests { auth ->
                auth.requestMatchers("/", "/index.html", "/static/**", "/css/**", "/js/**", "/favicon.ico").permitAll()
                    .requestMatchers("/login", "/signup", "/login.html", "/signup.html", "/api/v1/users/signup", "/api/v1/login", "/api/v1/refresh").permitAll()
                    .requestMatchers("/api/tasks/access-info").permitAll()
                    .requestMatchers("/h2-console/**").permitAll()
                    .requestMatchers("/actuator/health").permitAll()
                    .anyRequest().authenticated()
            }

            .addFilterBefore(jwtAuthenticationFilter, UsernamePasswordAuthenticationFilter::class.java)

            // 2. 커스텀 로그인 폼 활성화
            .formLogin { form ->
                form.loginPage("/login.html")
                    .loginProcessingUrl("/login")
                    .successHandler(formLoginSuccessHandler)
                    .failureHandler(formLoginFailureHandler)
                    .permitAll()
            }

            .oauth2Login { oauth2 ->
                oauth2.loginPage("/login.html")
                    .userInfoEndpoint { userInfo -> userInfo.userService(customOAuth2UserService) }
                    .successHandler(oAuth2SuccessHandler)
            }

            // 3. 예외 처리: API 요청에 대해서만 커스텀 핸들러(JSON) 적용
            .exceptionHandling { exception ->
                exception.defaultAuthenticationEntryPointFor(
                    authenticationHandler,
                    AntPathRequestMatcher("/api/**")
                )
            }
            
            .logout { logout ->
                logout.logoutUrl("/api/v1/logout")
                    .deleteCookies("accessToken", "refreshToken")
                    .logoutSuccessHandler(authenticationHandler)
            }
            .headers { headers -> headers.frameOptions { it.disable() } }

        return http.build()
    }

    @Bean
    @Throws(Exception::class)
    fun authenticationManager(authenticationConfiguration: AuthenticationConfiguration): AuthenticationManager {
        return authenticationConfiguration.authenticationManager
    }
}
